## **INTRODUCTION TO PYTHON FOR MACHNE LEARNING**

The Food Balance Sheets (FBS) dataset provides essential information on a country’s food system.

In [1]:
# package(s) needed
import pandas as pd

### **QUICK LOOK AT DATA**

In [2]:
df = pd.read_csv('FoodBalanceSheets_E_Africa_NOFLAG.csv', encoding='latin-1')
df.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

### **WHAT IS THE TOTAL NUMBER OF UNIQUE COUNTRIES IN THE DATASET?**

In [3]:
df['Area'].nunique()

49

### **PERFORM A GROUPBY OPERATION ON ‘ELEMENT’. WHAT YEAR HAS THE HIGHEST SUM OF STOCK VARIATION?**

In [4]:
df.groupby('Element')[['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']].sum()

Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein supply quantity (g/capita/day)    11842.45    11833.56  
Residuals                                 35500.00    34864.00  
Seed                                      24870.14    25263.14  
Stock Variation                           54316.91    20577.91  
Total Population - Both sexes           1112641.00  1140605.00  
Tourist consumption                          91.00       90.00

In [5]:
df.groupby('Element')[['Y2014', 'Y2015', 'Y2016', 'Y2017', 'Y2018']].sum()\
        .reset_index().query('Element == "Stock Variation"')

Element     Y2014     Y2015     Y2016     Y2017     Y2018
15  Stock Variation  58749.83  34910.99  33140.12  54316.91  20577.91

> The result above shows that the year 2014 has the highest sum of Stock Variation.

### **PERFORM A GROUPBY OPERATION ON ‘ELEMENT’. WHAT IS THE TOTAL NUMBER OF THE SUM OF PROCESSING IN 2017?**

In [6]:
df.groupby('Element')[['Y2017']].sum()

Y2017
Element                                           
Domestic supply quantity                2088198.10
Export Quantity                          182338.80
Fat supply quantity (g/capita/day)        10253.84
Feed                                     223705.68
Food                                    1258888.28
Food supply (kcal/capita/day)            454681.00
Food supply quantity (kg/capita/yr)       48690.04
Import Quantity                          294559.09
Losses                                   160614.00
Other uses (non-food)                     91645.97
Processing                               292836.00
Production                              2030056.89
Protein supply quantity (g/capita/day)    11842.45
Residuals                                 35500.00
Seed                                      24870.14
Stock Variation                           54316.91
Total Population - Both sexes           1112641.00
Tourist consumption                          91.00

In [7]:
df.groupby('Element')[['Y2017']].sum().reset_index().query('Element == "Processing"')

Element     Y2017
10  Processing  292836.0

> The total number of the sum of processing in 2017 is 292836.0.

### **SELECT COLUMNS ‘Y2017’ AND ‘AREA’, PERFORM A GROUPBY OPERATION ON ‘AREA’.  WHICH OF THESE AREAS HAD THE 7TH LOWEST SUM IN 2017?**

In [8]:
df.groupby('Area')[['Y2017', 'Area']].sum().sort_values(by='Y2017')[:10]

Y2017  \
Area                              
Sudan (former)             0.00   
Ethiopia PDR               0.00   
Comoros                   59.84   
Seychelles               442.34   
Sao Tome and Principe  12662.63   
Cabo Verde             14650.74   
Guinea-Bissau          19102.77   
Lesotho                21267.96   
Botswana               22101.30   
Djibouti               22729.91   

                                                                    Area  
Area                                                                      
Sudan (former)         Sudan (former)Sudan (former)Sudan (former)Suda...  
Ethiopia PDR           Ethiopia PDREthiopia PDREthiopia PDREthiopia P...  
Comoros                ComorosComorosComorosComorosComorosComorosComo...  
Seychelles             SeychellesSeychellesSeychellesSeychellesSeyche...  
Sao Tome and Principe  Sao Tome and PrincipeSao Tome and PrincipeSao ...  
Cabo Verde             Cabo VerdeCabo VerdeCabo VerdeCabo VerdeCabo V...  
Guinea-Bissau          Guinea-BissauGuinea-BissauGuinea-BissauGuinea-...  
Lesotho                LesothoLesothoLesothoLesothoLesothoLesothoLeso...  
Botswana               BotswanaBotswanaBotswanaBotswanaBotswanaBotswa...  
Djibouti               DjiboutiDjiboutiDjiboutiDjiboutiDjiboutiDjibou...

> Using the code above was concatenating the Area, so I used the code below to get a better format for the output.

In [9]:
df.groupby('Area')[['Y2017']].sum().sort_values(by='Y2017')[:10]

Y2017
Area                           
Sudan (former)             0.00
Ethiopia PDR               0.00
Comoros                   59.84
Seychelles               442.34
Sao Tome and Principe  12662.63
Cabo Verde             14650.74
Guinea-Bissau          19102.77
Lesotho                21267.96
Botswana               22101.30
Djibouti               22729.91

> The Area with the 7th lowest sum in year 2017 is Guinea-Bissau.

### **WHAT IS THE TOTAL SUM OF WINE PRODUCED IN 2015 AND 2018 RESPECTIVELY?**

In [10]:
df.groupby('Item')[['Y2015', 'Y2018']].sum()

Y2015      Y2018
Item                                     
Alcohol, Non-Food      2180.00    2293.00
Alcoholic Beverages   98783.72   97847.27
Animal Products       11811.73   11578.61
Animal fats          200675.72  269648.27
Apples and products   10559.15    9640.51
...                        ...        ...
Vegetables, Other    158104.08  163987.21
Vegetal Products     107064.17  107775.39
Wheat and products   234710.51  242645.19
Wine                   4251.81    4039.32
Yams                 203151.78  221272.09

[119 rows x 2 columns]

In [11]:
df.groupby('Item')[['Y2015', 'Y2018']].sum()\
    .reset_index().query('Item == "Wine"')

Item    Y2015    Y2018
117  Wine  4251.81  4039.32

> The total sum of Wine produced in 2015 and 2018 respectively are 4251.81 and 4039.32.

### **SELECT COLUMNS ‘Y2017’ AND ‘AREA’, PERFORM A GROUPBY OPERATION ON ‘AREA’.  WHICH OF THESE AREAS HAD THE HIGHEST SUM IN 2017?**

In [12]:
df.groupby('Area')[['Y2017', 'Area']].sum().sort_values(by='Y2017', ascending=False)[:10]

Y2017  \
Area                                      
Nigeria                      1483268.23   
Egypt                         866379.92   
South Africa                  517590.54   
Ethiopia                      448683.76   
Morocco                       388495.36   
Ghana                         337599.06   
Algeria                       325644.27   
United Republic of Tanzania   322616.85   
Kenya                         264660.66   
Sudan                         239931.92   

                                                                          Area  
Area                                                                            
Nigeria                      NigeriaNigeriaNigeriaNigeriaNigeriaNigeriaNige...  
Egypt                        EgyptEgyptEgyptEgyptEgyptEgyptEgyptEgyptEgyptE...  
South Africa                 South AfricaSouth AfricaSouth AfricaSouth Afri...  
Ethiopia                     EthiopiaEthiopiaEthiopiaEthiopiaEthiopiaEthiop...  
Morocco                      MoroccoMoroccoMoroccoMoroccoMoroccoMoroccoMoro...  
Ghana                        GhanaGhanaGhanaGhanaGhanaGhanaGhanaGhanaGhanaG...  
Algeria                      AlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlgeriaAlge...  
United Republic of Tanzania  United Republic of TanzaniaUnited Republic of ...  
Kenya                        KenyaKenyaKenyaKenyaKenyaKenyaKenyaKenyaKenyaK...  
Sudan                        SudanSudanSudanSudanSudanSudanSudanSudanSudanS...

> Using the code above was concatenating the Area, so I used the code below to get a better format for the output and I reduced the output to 10 rows.

In [13]:
df.groupby('Area')[['Y2017']].sum().sort_values(by='Y2017', ascending=False)[:10]

Y2017
Area                                   
Nigeria                      1483268.23
Egypt                         866379.92
South Africa                  517590.54
Ethiopia                      448683.76
Morocco                       388495.36
Ghana                         337599.06
Algeria                       325644.27
United Republic of Tanzania   322616.85
Kenya                         264660.66
Sudan                         239931.92

> The Area with the highest sum in 2017 is Nigeria.

### **WHAT IS THE MEAN AND STANDARD DEVIATION ACROSS THE WHOLE DATASET FOR THE YEAR 2017 TO 2 DECIMAL PLACES?**

In [14]:
df['Y2017'].describe()

count     59437.000000
mean        140.917765
std        1671.862359
min       -1582.000000
25%           0.000000
50%           0.100000
75%           9.000000
max      190873.000000
Name: Y2017, dtype: float64

> The mean and the standard deviation across the whole dataset for the year 2017 are 140.92 and 1671.86 respectively.

### **WHICH YEAR HAD THE LEAST CORRELATION WITH ‘ELEMENT CODE’?**

In [15]:
df.select_dtypes(include='number').corr()

Area Code  Item Code  Element Code     Y2014     Y2015  \
Area Code      1.000000  -0.005159     -0.000209  0.006164  0.005472   
Item Code     -0.005159   1.000000     -0.024683  0.021722  0.020857   
Element Code  -0.000209  -0.024683      1.000000  0.024457  0.023889   
Y2014          0.006164   0.021722      0.024457  1.000000  0.994647   
Y2015          0.005472   0.020857      0.023889  0.994647  1.000000   
Y2016          0.005247   0.020109      0.023444  0.996081  0.995739   
Y2017          0.005006   0.021494      0.024254  0.995230  0.988048   
Y2018          0.005665   0.021314      0.024279  0.994872  0.988208   

                 Y2016     Y2017     Y2018  
Area Code     0.005247  0.005006  0.005665  
Item Code     0.020109  0.021494  0.021314  
Element Code  0.023444  0.024254  0.024279  
Y2014         0.996081  0.995230  0.994872  
Y2015         0.995739  0.988048  0.988208  
Y2016         1.000000  0.992785  0.992757  
Y2017         0.992785  1.000000  0.998103  
Y2018         0.992757  0.998103  1.000000

> The year with the least correlation with Element Code is 2016.

### **WHAT IS THE TOTAL NUMBER AND PERCENTAGE OF MISSING DATA IN 2014 TO 3 DECIMAL PLACES?**

In [16]:
df['Y2014'].isnull().sum()

1589

In [17]:
round((df['Y2014'].isnull().sum() / df.shape[0]) * 100, 3)

2.607

> The total number and percentage of missing data in 2014 are 1589 and 2.607 3dp respectively.

### **WHAT IS THE TOTAL PROTEIN SUPPLY QUANTITY IN MADAGASCAR IN 2015?**

In [18]:
df.query('Element == "Protein supply quantity (g/capita/day)" and Area == "Madagascar"')\
    ['Y2015'].sum()

173.04999999999998

> The total protein supply quantity in Madagascar in 2015 is 173.05 to 3dp.